### PR TITLE
llama/ggml: Hadamard weight-fold runtime, ported to prism-v7

### PR DESCRIPTION
**Pull request by bri-prism**
Ports the Hadamard weight-fold runtime onto `prism-v7`. Supersedes #119 and #120, which are based on `prism-v6` and cannot be merged now that the README marks that branch do-not-build.

## What

Runtime support for models whose weights were Hadamard-folded offline, so a folded checkpoint can be packed to a low-bit format and still compute the original function:

- A `prism.hadamard` v1 GGUF contract (block size, transform, axis, sign mode, explicit sign vectors, per-tensor roles). The loader validates it, materializes one rotation tensor and one sign tensor per (width, buffer type), and applies the activation-side transform immediately before every folded matmul. A runtime that does not understand the contract refuses to load rather than silently skipping the transform.
- FWHT kernels extended from 512 to 1024 and 2048 wide, on CUDA and Metal, with F16 inputs accepted on CPU, CUDA, Metal and Vulkan. Backends without a matching kernel width fall back to a plain matmul against the real rotation matrix, so results stay correct everywhere.
- Inverse transform after the token-embedding lookup, so an embedding table can stay in the quantized latent format instead of a dequantized copy.
- Grouped V-head order for folded linear-attention out_proj, with the runtime permuting the activation instead: a post-fold column permutation on the rotation axis cannot be refolded without destroying the quantized codes.
- A graph-build check that every folded weight actually receives its transform, so an architecture whose matmul path bypasses the transform helper fails loudly instead of computing nonsense.
- `convert_hf_to_gguf` picks up a `hadamard_packing.json` manifest and writes the contract.

## Porting notes

Metal moved its shaders into `ggml/src/ggml-metal/kernels/` on this branch, so the FWHT kernel changes were re-landed in `kernels/misc.metal` rather than cherry-picked. Everything else applied as-is.

## Validation

- `test-backend-ops -o MUL_MAT_HADAMARD`: 18/18 on CUDA (RTX 4090, SM89, 2/2 backends) and 17/17 on Metal (3/3 backends), against the CPU reference, including the wide F32 and F16 cases and the fused sign-flip path at the three real contraction widths.
- End to end on a folded group-128 checkpoint: loads with 402 folded weights and 1 inverse-lookup table, and decodes at the same rate as on the previous branch (23.12 vs 23.16 tok/s).
- Numerics were certified on the previous branch and are unchanged by the port: logit KL divergence against the same model in unfolded F16 form sits at the measurement floor (mean 3.6e-4, same top-1 98.7%, PPL ratio statistically zero).

## One thing worth deciding

A file quantized before the `Q2_0_g128` to `PQ2_0` rename records ftype 142 while this branch uses 141, so it loads and computes correctly but prints `file type = unknown, may not work`. Anything quantized on this branch gets the right value automatically. If a legacy alias in the ftype name switch is wanted, say so and I will add it here; I left it out because the renumber looked deliberate.

### CODE REVIEW

File: ggml/src/ggml-cuda/ggml-cuda.cu [hunk 1]
```diff
@@ -3381,6 +3381,29 @@ static int ggml_cuda_try_fuse(ggml_backend_cuda_context * cuda_ctx, ggml_cgraph
...
+        ggml_tensor *       mm      = cgraph->nodes[i + 2];
+
+        const ggml_tensor * x     = mul->src[0];
+        const ggml_tensor * signs = mul->src[1];
+
+        const bool pattern_ok = ggml_get_op_params_i32(mm, 1) == GGML_HINT_SRC0_IS_HADAMARD &&
+            mm->src[1] == reshape && reshape->src[0] == mul &&
+            signs->ne[1] == 1 && signs->ne[2] == 1 && signs->ne[3] == 1 &&
+            signs->type == GGML_TYPE_F32 && mul->type == GGML_TYPE_F32 &&
+            (x->type == GGML_TYPE_F32 || x->type == GGML_TYPE_F16) &&
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply ggml/src/ggml-cuda/ggml-cuda.cu]: Confirmed and fixed in c0e4baf2b. ggml_mul_impl builds its result with ggml_dup_tensor(ctx, a), so mul->type follows x->type, and the gate required mul->type == F32 while allowing x to be F16. The whole condition collapsed to F32 only, which means the templated signed half kernel was dead and the F16 signed test was covering the unfused path exactly as you describe. The gate now requires mul->type == x->type with either supported input type. One caveat: this is the CUDA graph gate and I have no NVIDIA host right now, so it is not covered by the 17/17 MUL_MAT_HADAMARD run I did on Metal. It needs a CUDA run before merge.

File: src/llama-model.cpp [hunk 2]
```diff
@@ -1189,6 +1191,141 @@ void llama_model_base::load_hparams(llama_model_loader & ml) {
...
+        }
+        if (weight_names.empty()) {
+            throw std::runtime_error("prism.hadamard.weight_names is empty");
+        }
+
+        if (sign_mode == "explicit") {
+            std::vector<int32_t> sign_widths;
+            std::vector<int32_t> sign_values;
+            ml.get_arr("prism.hadamard.sign_widths", sign_widths);
+            ml.get_arr("prism.hadamard.sign_values", sign_values);
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply src/llama-model.cpp:1232]: Correct, fixed in c0e4baf2b. With sign_widths empty the loop never runs, off stays 0, and the trailing off != sign_values.size() check compares 0 against 0 and passes, so the contract validates and the sign table stays empty. Explicit mode now requires at least one width.

File: conversion/base.py [hunk 3]
```diff
@@ -617,6 +617,139 @@ def map_tensor_name(self, name: str, try_suffixes: Sequence[str] = (".weight", "
...
+        sign_widths: list[int] = []
+        sign_values: list[int] = []
+        if sign_mode == "explicit":
+            signs = manifest.get("signs")
+            if not isinstance(signs, dict) or not signs:
+                raise ValueError("explicit sign mode requires a signs table")
+            for width_str, vec in sorted(signs.items(), key=lambda kv: int(kv[0])):
+                width = int(width_str)
+                if len(vec) != width or any(v not in (-1, 1) for v in vec):
+                    raise ValueError(f"invalid sign vector for width {width}")
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/base.py:675]: Fixed in c0e4baf2b. The converter now applies the same rule as src/llama-model.cpp, rejecting a width that is not positive or not a multiple of block_size, so a manifest that converts also loads.

File: conversion/qwen.py [hunk 4]
```diff
@@ -605,8 +605,18 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
                 data_torch = torch.cat([qk_part, v_part], dim=0)
 
             elif ".out_proj." in name:
-                # Out projection weight: reorder columns (input dimension)
-                data_torch = self._reorder_v_heads(data_torch, 1, num_k_heads, num_v_per_k, head_v_dim)
+                # the tensor name may have had wrapper prefixes stripped by the
+                # time it reaches modify_tensors, so match manifest entries by
+                # suffix rather than exact name
+                suffix = "linear_attn.out_proj.weight"
+                if name.endswith(suffix) and any(n.endswith(suffix) for n in self.hadamard_folded_names()):
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/qwen.py]: Fixed in c0e4baf2b. The suffix test is replaced with a per-tensor match that compares the full name modulo leading wrapper prefixes, so layer identity is preserved and a partially folded checkpoint no longer puts every layer's out_proj in grouped order.

File: conversion/qwen.py [hunk 5]
```diff
@@ -605,8 +605,18 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
...
+                    # Hadamard-folded latent: the rotation axis must keep the
+                    # training (grouped) V order; the runtime permutes the
+                    # activation tiled->grouped before the transform instead.
+                    self._hadamard_gdn_v_grouped = True
```
[the remark this answers is not in the document]
Comment (bri-prism) [reply conversion/qwen.py:626]: Good catch, and this was the worst of the five since it produces wrong numbers rather than a load failure. Fixed in c0e4baf2b: the out_proj branch in _transform_nvfp4_weight now takes the same guard as modify_tensors, so a folded out-projection keeps grouped order and sets the grouped flag instead of being column-permuted on the rotation axis. Both paths now go through one helper, which is what let them drift apart in the first place. Same caveat as the CUDA gate: I have not run an NVFP4 prepacked conversion end to end, so this is verified by reading rather than by output.
